mls-rs/src/group/mod.rs

```diff
@@ -908,32 +941,84 @@ where
         signer: Option<SignatureSecretKey>,
         signing_identity: Option<SigningIdentity>,
     ) -> Result<Proposal, MlsError> {
+        let leaf_node = self.updated_leaf_node(signer, signing_identity).await?;
+        Ok(Proposal::Update(UpdateProposal { leaf_node }))
+    }
+
+    /// Create a proposal message that replaces another member.
+    ///
+    /// `authenticated_data` will be sent unencrypted along with the contents
+    /// of the proposal message.
+    #[cfg(feature = "replace_proposal")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn propose_replace(
+        &mut self,
+        to_replace: u32,
+        leaf_node: LeafNode,
+        authenticated_data: Vec<u8>,
+    ) -> Result<MlsMessage, MlsError> {
+        let proposal = self.replace_proposal(to_replace, leaf_node).await?;
+        self.proposal_message(proposal, authenticated_data).await
+    }
+
+    #[cfg(feature = "replace_proposal")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn replace_proposal(
+        &mut self,
+        to_replace: u32,
+        leaf_node: LeafNode,
+    ) -> Result<Proposal, MlsError> {
+        Ok(Proposal::Replace(ReplaceProposal {
+            to_replace: LeafIndex(to_replace),
+            leaf_node,
+        }))
+    }
+
+    /// Abandon any cached state corresponding to a leaf node
+    #[cfg(feature = "replace_proposal")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn abandon_leaf_node(&mut self, leaf_node: &LeafNode) {
+        self.pending_updates.remove(&leaf_node.public_key);
+    }
```

Comment:
Isn't it annoying for the application to track  `LeafNode`'s?

Can we make it automatic so that leaves with epoch <= current leaf epoch are deleted automatically since they can no longer be applied? Storing other leaves can't hurt security (since they haven't been applied) so this function isn't so crucial anymore.

Also, how about a function that deletes all cached leaves, or all leaves with `not_before < given time T`, or all leaves from epoch `< given epoch E`?